File: app/src/main/java/net/schueller/peertube/helper/VideoHelper.java
/*
 * Copyright (C) 2020 Stefan Schüller <sschueller@techdroid.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package net.schueller.peertube.helper;

import android.app.AppOpsManager;
import android.content.Context;
import android.content.SharedPreferences;
import android.os.Build;
import android.preference.PreferenceManager;
import android.util.Log;

import net.schueller.peertube.R;

public class VideoHelper {

    private static final String TAG = "VideoHelper";

    public static boolean canEnterPipMode(Context context) {

        SharedPreferences sharedPref = PreferenceManager.getDefaultSharedPreferences(context);

        context.getString(R.string.pref_background_float_key);

        // pref is disabled
        if (!context.getString(R.string.pref_background_float_key).equals(
                sharedPref.getString(
                        context.getString(R.string.pref_background_behavior_key),
                        context.getString(R.string.pref_background_float_key))
            )
        ) {
            return false;
        }

        // api does not support it
        Log.v(TAG, "api version " + Build.VERSION.SDK_INT);
        if (Build.VERSION.SDK_INT > 27) {
            AppOpsManager appOpsManager = (AppOpsManager) context.getSystemService(Context.APP_OPS_SERVICE);
            return (AppOpsManager.MODE_ALLOWED == appOpsManager.checkOp(AppOpsManager.OPSTR_PICTURE_IN_PICTURE, android.os.Process.myUid(), context.getPackageName()));
        }
        return false;
    }
}

File: app/src/main/java/net/schueller/peertube/common/VideoHelper.kt
package net.schueller.peertube.common

import net.schueller.peertube.feature_video.domain.model.Video
import javax.inject.Inject
import javax.inject.Singleton
import android.app.AppOpsManager

import android.os.Build

import android.R
import android.content.Context

import android.preference.PreferenceManager

import android.content.SharedPreferences
import android.os.Process
import dagger.hilt.android.qualifiers.ApplicationContext
import net.schueller.peertube.common.Constants.PREF_BACKGROUND_BEHAVIOR_KEY
import net.schueller.peertube.common.Constants.PREF_BACKGROUND_FLOAT_KEY


@Singleton
class VideoHelper @Inject constructor(
) {



    fun pickPlaybackResolution(video: Video, preferredQuality: Int = 999999): String?
    {
        var urlToPlay: String? = null

        if (video.streamingPlaylists != null && video.streamingPlaylists.isNotEmpty()) {
            urlToPlay = video.streamingPlaylists[0].playlistUrl
        } else {
            if (video.files != null && video.files.isNotEmpty()) {
                urlToPlay = video.files[0].fileUrl // default, take first found, usually highest res
                for (file in video.files) {
                    // Set quality if it matches
                    if (file.resolution.id == preferredQuality) {
                        urlToPlay = file.fileUrl
                    }
                }
            }
        }

        return urlToPlay
    }


    fun canEnterPipMode(context: Context): Boolean {
        val sharedPreferences = context.getSharedPreferences(context.packageName + "_preferences", Context.MODE_PRIVATE)

        // pref is disabled
        if (!PREF_BACKGROUND_FLOAT_KEY.equals(
                sharedPreferences.getString(
                    PREF_BACKGROUND_BEHAVIOR_KEY,
                    PREF_BACKGROUND_FLOAT_KEY
                )
            )
        ) {
            return false
        }

        // api does not support it
        if (Build.VERSION.SDK_INT > 27) {
            val appOpsManager = context.getSystemService(Context.APP_OPS_SERVICE) as AppOpsManager
            return AppOpsManager.MODE_ALLOWED == appOpsManager.checkOp(
                AppOpsManager.OPSTR_PICTURE_IN_PICTURE,
                Process.myUid(),
                context.packageName
            )
        }
        return false
    }

}

